Replace `minimp3` with `dr_mp3`

DeeJayLSP opened the pull request:
Last time `minimp3` got updated was in November 2021 (almost 3 years ago by the time this PR was opened).

Meanwhile, `dr_mp3`, a piece of work done on top of `minimp3`, has been updated quite recently.

Within Godot, it fixes a few issues:
- No need for patches;
- No more MSVC warnings (removed system header treatment);
- Possibility of overriding the memory allocator to use Godot's own.
- No more link-time optimization warnings (supersedes #109788).

Template release binary sizes remain the same.

Since `minimp3` is removed, the following are also changed in this PR:
- `minimp3` module renamed to `mp3`, to better reflect its purpose rather than naming its dependency;
- `minimp3_extra_formats` build variable renamed to `mp3_extra_formats`.

This should not break compatibility unless module/build option name changes count (does not affect existing projects™).

DeeJayLSP commented:
It seems like the change adds a 0.05s silence at the beginning, gonna check if there's a way to fix this

DeeJayLSP commented:
Current version has the issue fixed. Despite being a preview version I didn't notice any problems.

Maybe it's safe to mark as ready now.

DeeJayLSP commented:
Considering that I'm also trying to make a module named `dr_wav`, should the module in this PR be named `dr_mp3` instead of just `mp3`?

Calinou commented:
I would rename the module to just `mp3`. It's more intuitive for people using SCons options to disable features, and it follows the naming scheme of most existing modules more closely (`gltf`, `fbx`).

DeeJayLSP commented:
Rebased on top of master. Added a `dr_alloc_calls.h` file also used by `dr_wav` in #96545.

Due to renaming an entire module folder, any minor changes to the `minimp3` module require a rebase.

DeeJayLSP commented:
I was trying to for #96545 to be approved first, as both PRs share files from `dr_libs`, but it seems there are some problems with `minimp3` that don't happen with `dr_mp3`, so I'm going to give more priority to this one.

Calinou reviewed (approved):
Tested locally on the [Run-time saving and loading demo](https://github.com/godotengine/godot-demo-projects/tree/master/loading/runtime_save_load), it works as expected. Code looks good to me.

I can confirm building with LTO also no longer reproduces warnings for MP3 code on GCC 15.2.1. There are still plenty of other LTO warnings produced though (e.g. by polypartition, VHACD, libjpeg-turbo, …).

**Edit:** There are some bugs with the PR, see https://github.com/godotengine/godot/pull/96547#issuecomment-3353073875.

DeeJayLSP commented:
Rebased.

> I actually found that MP3 playback breaks on some longer files (resulting in audible stuttering and slowed down playback), even though it appears fine in the first 5 seconds of playback. It plays correctly in `master`.

I tried to modify the PR so the playback procedure would be similar to the Ogg Vorbis player. Reverting that seems to have solved this issue.

DeeJayLSP committed:
Replace `minimp3` with `dr_mp3`

DeeJayLSP commented:
Rebased after recent changes.

adamscott reviewed (approved):
Didn't test any MP3 files, but the code seems right. Thanks @DeeJayLSP !

DeeJayLSP commented on COPYRIGHT.txt:
```diff
@@ -253,6 +253,11 @@ Comment: doctest
+Files: thirdparty/dr_libs/*
+Comment: dr_libs
+Copyright: 2020 David Reid
+License: public-domain or Unlicense or Expat
```
It looks like the [miniaudio Debian copyright file](https://metadata.ftp-master.debian.org/changelogs//main/m/miniaudio/miniaudio_0.11.22+dfsg-1_copyright) actually uses "MIT-0". Should we go with it? Sorry I only came up with the idea of looking examples too late.

akien-mga commented on thirdparty/README.md:
```diff
@@ -200,6 +200,20 @@ Patches:
+## dr_libs
+
+- Upstream: https://github.com/mackron/dr_libs
+- Version: mp3-0.7.1 (9de1c5dff6810cdb1a9c69a97a9808495168c4d0, 2025)
```
There seems to be `mp3-0.7.2` now.

akien-mga reviewed (approved):
Looks good to me, apologies for the delay.

Since this is mainly swapping a thirdparty dependency for another compatible one, as thirdparty maintainer I'll allow this one to slip past yesterday's feature freeze and include it in 4.6.

DeeJayLSP commented:
Updated from `mp3-0.7.1` to `mp3-0.7.2`.

Being a patch version and checking the diff, I don't think it's harmful.

The [force-push diff](https://github.com/godotengine/godot/compare/fdd25f7000b05db1a0e45c8675cc5c8f3cf20f95..f46bca99c743630693874de7c94c3334dc54c8f4) is quite simple.

Repiteo commented:
Thanks!
